docs: Add Signals tutorial with examples and hints in guide (#1336)
preactjs/preact-www: content/en/tutorial/10-signals.md

---
title: Signals
prev: /tutorial/09-error-handling
next: /tutorial/11-links
solvable: true
---

> # Signals

Reviewer: Not essential, but some more consistent line length in this document would be appreciated. Some lines wrap early, some go long; it's just a bit inconsistent when editing.

> In chapter four, we saw how Preact manages state with the help of
> **hooks**. Hooks are powerful, but they always re-render the whole component whenever the state changes.
> 

Reviewer: ```suggestion
In chapter four, we saw how Preact manages state with the help of **hooks**. Hooks
rely upon re-rendering the component on state/prop change, and whilst this is a perfectly
usable set of APIs, we find many users struggle with modeling their components and using
this system without consistently running into some known pitfalls.
```

> **Signals** is a other way to manage state in Preact. Unlike hooks, which re-render the whole component whenever the state changes, the signals update only the
> parts of the UI that actually uses them. A signal remembers its value and automatically tracks where it is used, so when the value
> changes, only the parts that depend on it will update.

Reviewer: ```suggestion
**Signals** are another way to manage state in Preact. Signals can offer automatic fine-grained
updates, avoiding the need for dependency arrays and skipping full-component re-renders, as
well as a (subjectively) simpler flow for derived state. This helps ensure apps of all sizes are fast by default.
```

> ### Creating a signal

Reviewer: I feel like a lot of this content should be copy/pasted from the Signals guide page or intro blog. It looks like you wrote this entirely yourself, which is impressive, but the language and examples diverge quite a bit despite containing similar information.

Don't have time to review more right now, but I can take a look at this myself at some point.

}
```

Here, `counter` and `doubled` are global signals that both components (Increment and Display) can use directly.

```jsx
import { signal, computed } from '@preact/signals';
import { createContext } from 'preact';
import { useContext } from 'preact/hooks';

// Create a state object
function createAppState() {
	const counter = signal(0);
	const doubled = computed(() => counter.value * 2);
	return { counter, doubled };
}

// Create a context
const AppState = createContext();

// Component 1: Increment button
function Increment() {
	const state = useContext(AppState); // access shared state
	return <button onClick={() => state.counter.value++}>Increment</button>;
}

// Component 2: Display result
function Display() {
	const state = useContext(AppState); // access shared state
	return <p>Counter doubled: {state.doubled}</p>;
}

// Main App
export function App() {
	return (
		<AppState.Provider value={createAppState()}>
			<Increment />
			<Display />
		</AppState.Provider>
	);
}
```

Here, `counter` and `doubled` is created once inside `createAppState()` and passed to all components through `Context`,
so you don’t need to import global signals directly.

## Try it!

> Let’s create a counter which gets incremented while the button is clicked using `signal`, which is similar to the state example from fourth chapter.

Reviewer: We can drop the mention of the fourth chapter, but I'd ask that we add a mention that the user should display the current count (`{count}`) -- I realize chapter 4 doesn't really mention that either, but that's a mistake & sort of confusing, IMO.

</solution>

```js:setup
useResult(function() {
	var options = require('preact').options;

	var oe = options.event;
	options.event = function(e) {
		if (oe) oe.apply(this, arguments);

		if (e.currentTarget.localName !== 'button') return;
		var root = e.currentTarget.parentNode.parentNode;
		var text = root.innerText.match(/Count:\s*([\w.-]*)/i);
		if (!text) return;
		if (!text[1].match(/^-?\d+$/)) {
			return console.warn(
				"Tip: it looks like you're not rendering {count} anywhere."
			);
		}
		setTimeout(function() {
			var text2 = root.innerText.match(/Count:\s*([\w.-]*)/i);
			if (!text2) {
				return console.warn('Tip: did you remember to render {count}?');
			}
			if (text2[1] == text[1]) {
				return console.warn(
					'Tip: remember to update the signal value using `.value`.'
				);
			}
			if (!text2[1].match(/^-?\d+$/)) {
				return console.warn(
					'Tip: it looks like `count` is being set to something other than a number.'
				);
			}

			if (Number(text2[1]) === Number(text[1]) + 1) {
				solutionCtx.setSolved(true);
			}
		}, 10);
	};

	return function() {
		options.event = oe;
	};
}, []);
```

```jsx:repl-initial
import { render } from 'preact';

> import { signal } from '@preact/signals';

Reviewer: ```suggestion
import { useSignal } from '@preact/signals';
```

function MyButton(props) {
	return (
		<button style={props.style} onClick={props.onClick}>
			{props.children}
		</button>
	);
}

function App() {

> 	const count = signal(0);
> 

Reviewer: Let them create the signal, matching 04-state

```suggestion
```

	const clicked = () => {
		// increment the signal value here
	};

	return (
		<div>
			<p class="count">Count:</p>
			<MyButton style={{ color: 'purple' }} onClick={clicked}>
				Click me
			</MyButton>
		</div>
	);
}

render(<App />, document.getElementById('app'));
```

```jsx:repl-final
import { render } from 'preact';

> import { signal } from '@preact/signals';

Reviewer: ```suggestion
import { useSignal } from '@preact/signals';
```

function MyButton(props) {
	return (
		<button style={props.style} onClick={props.onClick}>
			{props.children}
		</button>
	);
}

function App() {

> 	const count = signal(0);

Reviewer: ```suggestion
	const count = useSignal(0);
```
